File: backend/src/main/java/br/ufpr/sept/so2/modules/academico/infrastructure/persistence/PeriodoLetivoJpaAdapter.java
package br.ufpr.sept.so2.modules.academico.infrastructure.persistence;

import br.ufpr.sept.so2.modules.academico.application.ports.PeriodoLetivoRepository;
import br.ufpr.sept.so2.modules.academico.domain.PeriodoLetivo;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Component
public class PeriodoLetivoJpaAdapter implements PeriodoLetivoRepository {

    private final PeriodoLetivoJpaRepository jpaRepository;

    public PeriodoLetivoJpaAdapter(PeriodoLetivoJpaRepository jpaRepository) {
        this.jpaRepository = jpaRepository;
    }

    @Override
    public PeriodoLetivo save(PeriodoLetivo periodo) {
        PeriodoLetivoJpaEntity entity = jpaRepository.findById(periodo.getId())
                .orElseGet(() -> PeriodoLetivoJpaEntity.fromDomain(periodo));
        entity.merge(periodo);
        return jpaRepository.save(entity).toDomain();
    }

    @Override
    public Optional<PeriodoLetivo> findById(UUID id) {
        return jpaRepository.findById(id).map(PeriodoLetivoJpaEntity::toDomain);
    }

    @Override
    public Page<PeriodoLetivo> findAll(Pageable pageable) {
        return jpaRepository.findAll(pageable).map(PeriodoLetivoJpaEntity::toDomain);
    }

    @Override
    public List<PeriodoLetivo> findAll() {
        return jpaRepository.findAll().stream().map(PeriodoLetivoJpaEntity::toDomain).toList();
    }

    @Override
    public Optional<PeriodoLetivo> findVigente(LocalDate data) {
        return jpaRepository.findVigente(data).map(PeriodoLetivoJpaEntity::toDomain);
    }

    @Override
    public boolean existsByAnoAndSemestre(int ano, int semestre) {
        return jpaRepository.existsByAnoAndSemestre((short) ano, (short) semestre);
    }

    @Override
    public void deleteById(UUID id) {
        jpaRepository.deleteById(id);
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/academico/infrastructure/persistence/PeriodoLetivoJpaAdapter.kt
package br.ufpr.sept.so2.modules.academico.infrastructure.persistence

import br.ufpr.sept.so2.modules.academico.application.ports.PeriodoLetivoRepository
import br.ufpr.sept.so2.modules.academico.domain.PeriodoLetivo
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component
import java.time.LocalDate
import java.util.Optional
import java.util.UUID

@Component
class PeriodoLetivoJpaAdapter(
    private val jpaRepository: PeriodoLetivoJpaRepository,
) : PeriodoLetivoRepository {
    override fun save(periodo: PeriodoLetivo): PeriodoLetivo {
        val entity = jpaRepository.findById(periodo.id).orElseGet { PeriodoLetivoJpaEntity.fromDomain(periodo) }
        entity.merge(periodo)
        return jpaRepository.save(entity).toDomain()
    }

    override fun findById(id: UUID): Optional<PeriodoLetivo> =
        jpaRepository.findById(id).map { it.toDomain() }

    override fun findAll(pageable: Pageable): Page<PeriodoLetivo> =
        jpaRepository.findAll(pageable).map { it.toDomain() }

    override fun findAll(): List<PeriodoLetivo> =
        jpaRepository.findAll().map { it.toDomain() }

    override fun findVigente(data: LocalDate): Optional<PeriodoLetivo> =
        jpaRepository.findVigente(data).map { it.toDomain() }

    override fun existsByAnoAndSemestre(ano: Int, semestre: Int): Boolean =
        jpaRepository.existsByAnoAndSemestre(ano.toShort(), semestre.toShort())

    override fun deleteById(id: UUID) {
        jpaRepository.deleteById(id)
    }
}
